### PR TITLE
don't link substitutions, require them for as_posyslt1()

### LINKED ISSUE #827: unknown solution status dependent on constraint list ordering
I just ran into some behavior I really don't understand. These are the constraints for my combustor model, and everything works fine with them. 

```python
           constraints = [
                #flow through combustor
                Pt4 == pib * Pt3,   #B.145
                ht4 == Cpc * Tt4,
                
                TCS([f*hf >= (1-ac)*ht4-(1-ac)*ht3+Cpfuel*f*(Tt4-Ttf)]),

                #making f+1 GP compatible --> needed for convergence
                SignomialEquality(fp1,f+1),
                
                ht41 == Cpc * Tt41,
                ]
            
            constraints.extend([
                #compute Tt41...mixing causes a temperature drop
                #had to include Tt4 here to prevent it from being pushed down to zero
                TCS([ht41 <= ((1-ac)*ht4 +ac*ht3+Cpfuel*f*(Tt4-Ttf) + Cpfuel*Ttf*f)/fp1]),
               
                #comptue the rest of the station 4.1 variables
                SignomialEquality(fp1*u41, (u4a*(1-ac)+f*u4a+ac*uc)),          
                
                #this is a stagnation relation...need to fix it to not be signomial
                TCS([T41 >= Tt41-.5*(u41**2)/Cpc]),
                #here we assume no pressure loss in mixing so P41=P4a
                Pt41 == P4a*(Tt41/T41)**(1.313/.313),
                #compute station 4a quantities, assumes a gamma value of 1.313 (air @ 1400K)
                u4a == M4a*((1.313*R*Tt4)**.5)/hold4a,
                uc == ruc*u4a,
                P4a == Pt4*hold4a**(-1.313/.313),
                ])
```

however, these constraints cause the model to return unknown

```python
           constraints = [
                #flow through combustor
                Pt4 == pib * Pt3,   #B.145
                ht4 == Cpc * Tt4,

                #making f+1 GP compatible --> needed for convergence
                SignomialEquality(fp1,f+1),
                
                ht41 == Cpc * Tt41,
                ]
            
            constraints.extend([
               TCS([f*hf >= (1-ac)*ht4-(1-ac)*ht3+Cpfuel*f*(Tt4-Ttf)]),
                #compute Tt41...mixing causes a temperature drop
                #had to include Tt4 here to prevent it from being pushed down to zero
                TCS([ht41 <= ((1-ac)*ht4 +ac*ht3+Cpfuel*f*(Tt4-Ttf) + Cpfuel*Ttf*f)/fp1]),
               
                #comptue the rest of the station 4.1 variables
                SignomialEquality(fp1*u41, (u4a*(1-ac)+f*u4a+ac*uc)),          
                
                #this is a stagnation relation...need to fix it to not be signomial
                TCS([T41 >= Tt41-.5*(u41**2)/Cpc]),
                #here we assume no pressure loss in mixing so P41=P4a
                Pt41 == P4a*(Tt41/T41)**(1.313/.313),
                #compute station 4a quantities, assumes a gamma value of 1.313 (air @ 1400K)
                u4a == M4a*((1.313*R*Tt4)**.5)/hold4a,
                uc == ruc*u4a,
                P4a == Pt4*hold4a**(-1.313/.313),
                ])
```

the only difference is that this constraint ``TCS([f*hf >= (1-ac)*ht4-(1-ac)*ht3+Cpfuel*f*(Tt4-Ttf)]),`` was moved from the constraints = [] to constraints.extend([])...any idea what's going on here?

### PR DESCRIPTION
**Pull request by bqpd**
I decided that substitutions actually shouldn't be linked, and this branch implements that in the proper fashion. Should close #827, removes some ugly lines of code...

Oh, and it fixes an unrelated bug @mjburton11 found today as well.

@galbramc, there's some pylint errors in this that I couldn't get to disappear. The no-member one is silly, since it's doing isinstance on `ConstraintSet`, not `list`, but I couldn't get it to go away with a comment. The circular import one is less silly (it's a mild hack) but I can't get it to go away either.

### CODE REVIEW

**Comment by galbramc** [comment]
I was able to suppress the no-member warning with # pylint: disable=no-member. What version of pylint do you have on your computer? Jenkins is using 1.6.4 now, and that's what I have on mine.

As far as the cyclic import, that is really a bad sign. Its frail and is likely easily broken as you continue to work on the code. I would strongly urge you to try to find a way to refactor the code to avoid the cyclic import. @whoburg. it's ultimately your call, but this cyclic import is likely to break sooner than later.

**Comment by bqpd** [comment]
If you look at where it is in the code, though, it's really more of a wrong import order; it gets called in a function call, so it's safe from the breaking nature of cyclic imports but just kinda ugly.

**Comment by bqpd** [comment]
err, why didn't this run the unit tests?

**Comment by galbramc** [comment]
thats odd

test this please

**Comment by galbramc** [comment]
No idea why the unit tests didn't run. 

There is no way of cleaning up the cyclic import? Its easier to do now while you have things fresh in your mind.

**Comment by bqpd** [comment]
Updated pylint. Unfortunately there isn't an easy one. We've had the similar `from ..nomials import Variable` in that file for a bit for the same reason.

**Comment by bqpd** [comment]
@whoburg went a little late-night overboard on the repo cleanup, so I recomend looking at the commits individually...ready for review

**Comment by bqpd** [comment]
@galbramc, and re cylic imports, it's [reccomended](http://stackoverflow.com/questions/7336802/how-to-avoid-circular-imports-in-python) to use absolute imports as the way around this: that is, `import gpkit.Variable` instead of doing relative imports. I'd argue that is an even hackier solution.

(in Python 3 it doesn't have this issue and you can just put the imports at the top of the file)

**Comment by galbramc** [comment]
Were you able to disable pylint with an inline comment for the cyclic import?

**Comment by bqpd** [comment]
No, sadly.

**Comment by galbramc** [comment]
You were able to fix the pylint warning somehow?

**Comment by bqpd** [comment]
I have no idea why it disappeared...

**Comment by whoburg** [comment]
retest this please

File: gpkit/constraints/set.py [hunk 1]
```diff
@@ -49,31 +54,32 @@ def __init__(self, constraints, substitutions=None):
                 # TODO: remove conditional or otherwise clean up
                 if hasattr(self[i], "substitutions"):
                     self.substitutions.update(self[i].substitutions)
-        self.substitutions.update(subs)
         self.reset_varkeys()
-        # initializations for attributes used elsewhere
-        self.posymap = []
+        self.substitutions.update({self[k].key: v for k, v in subs.items()
+                                   if k in self.varkeys})
+        # TODO: on all updates the keydict should do the referencing above
```
Comment (whoburg) [inline gpkit/constraints/set.py]: concerned about speed here

File: gpkit/solution_array.py [hunk 2]
```diff
@@ -131,7 +131,7 @@ def table(self, tables=("cost", "sweepvariables", "freevariables",
                 if not subdict["constants"]:
                     continue
                 strs += results_table(subdict["constants"], table_title,
-                                      minval=1e-2,
+                                      minval=0.1,
```
Comment (whoburg) [inline gpkit/solution_array.py]: I prefer refactoring `minval` to print the top N sensitivities.

**Comment by whoburg** [comment]
ok to merge after passing tests and addressing potential speed issue on model[varkey] lookup

**Pull request merged**